File: app/services/knowledge_service.py
from __future__ import annotations

from typing import Any

import requests

from app.config import get_settings


class KnowledgeService:
    def __init__(self) -> None:
        self.settings = get_settings()

    def search(self, query: str, top_k: int = 5) -> list[dict[str, Any]]:
        if not self.settings.knowledge_search_url:
            return []

        response = requests.get(
            self.settings.knowledge_search_url,
            params={"q": query, "top_k": top_k},
            timeout=self.settings.knowledge_search_timeout_sec,
        )
        response.raise_for_status()
        payload = response.json()

        if isinstance(payload, list):
            return payload
        if isinstance(payload, dict):
            for key in ("items", "data", "results"):
                value = payload.get(key)
                if isinstance(value, list):
                    return value
        return []

File: src/main/java/com/gczm/aimeeting/service/KnowledgeService.java
package com.gczm.aimeeting.service;

import com.gczm.aimeeting.config.AppProperties;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import org.apache.commons.lang3.StringUtils;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.util.UriComponentsBuilder;

import java.util.List;
import java.util.Map;

@Service
@RequiredArgsConstructor
public class KnowledgeService {

    private final AppProperties appProperties;
    private final RestTemplate restTemplate;
    private final ObjectMapper objectMapper;

    public List<Map<String, Object>> search(String query, int topK) {
        String url = appProperties.getKnowledge().getSearchUrl();
        if (StringUtils.isBlank(url)) {
            return List.of();
        }

        String finalUrl = UriComponentsBuilder.fromHttpUrl(url)
                .queryParam("q", query)
                .queryParam("top_k", topK)
                .toUriString();

        ResponseEntity<String> response = restTemplate.getForEntity(finalUrl, String.class);
        if (!response.getStatusCode().is2xxSuccessful() || response.getBody() == null) {
            return List.of();
        }

        try {
            Object payload = objectMapper.readValue(response.getBody(), Object.class);
            if (payload instanceof List<?> list) {
                return list.stream().filter(Map.class::isInstance).map(m -> (Map<String, Object>) m).toList();
            }
            if (payload instanceof Map<?, ?> map) {
                for (String key : List.of("items", "data", "results")) {
                    Object value = map.get(key);
                    if (value instanceof List<?> list) {
                        return list.stream().filter(Map.class::isInstance).map(m -> (Map<String, Object>) m).toList();
                    }
                }
            }
        } catch (Exception ignored) {
            return List.of();
        }

        return List.of();
    }
}
